pkg/scheduler/api/hyper_node_info.go

```diff
@@ -0,0 +1,463 @@
+			if !ancestors.Has(parent) {
+				ancestors.Insert(parent)
+				queue = append(queue, parent)
+			}
+		}
+	}
+	return ancestors
+}
+
+// GetDescendants returns all descendants of a given HyperNode.
+func (hni *HyperNodesInfo) GetDescendants(hyperNodeName string) sets.Set[string] {
+	queue := []string{hyperNodeName}
+	descendants := sets.New[string]()
+
+	for len(queue) > 0 {
+		current := queue[0]
+		queue = queue[1:]
+		descendants.Insert(current)
+		if hn, ok := hni.hyperNodes[current]; ok {
+			for _, member := range hn.HyperNode.Spec.Members {
+				if member.Type == topologyv1alpha1.MemberTypeHyperNode && member.Selector.ExactMatch != nil {
+					childName := member.Selector.ExactMatch.Name
+					if !descendants.Has(childName) {
+						queue = append(queue, childName)
+					}
+				}
+			}
+		}
+	}
+	return descendants
+}
+
+// getChildren returns the immediate children of a given HyperNode.
+func (hni *HyperNodesInfo) getChildren(hyperNodeName string) sets.Set[string] {
+	children := sets.New[string]()
+	hn, ok := hni.hyperNodes[hyperNodeName]
+	if !ok {
+		klog.ErrorS(nil, "HyperNode not found", "name", hyperNodeName)
+		return children
+	}
+
+	for _, member := range hn.HyperNode.Spec.Members {
+		if member.Type == topologyv1alpha1.MemberTypeHyperNode && member.Selector.ExactMatch != nil {
+			childName := member.Selector.ExactMatch.Name
+			children.Insert(childName)
+		}
+	}
+	return children
+}
+
+// setParent sets the parent of a HyperNode member.
+func (hni *HyperNodesInfo) setParent(hn *HyperNodeInfo, memberName string) error {
+	parent, ok := hni.parentMap[memberName]
+	if ok && parent != hn.Name {
+		return fmt.Errorf("HyperNode %s already has a parent %s, and cannot set another parent %s", memberName, parent, hn.Name)
+	} else {
+		hni.parentMap[memberName] = hn.Name
+	}
+	return nil
+}
+
+// getMembers retrieves the members of a HyperNode based on the selector.
+func (hni *HyperNodesInfo) getMembers(selector topologyv1alpha1.MemberSelector, nodes []*corev1.Node) sets.Set[string] {
+	if selector.ExactMatch != nil {
+		if selector.ExactMatch.Name == "" {
+			return sets.New[string]()
+		}
+		return sets.New[string](selector.ExactMatch.Name)
+	}
+
+	members := sets.New[string]()
+	if selector.RegexMatch != nil {
+		pattern := selector.RegexMatch.Pattern
+		reg, err := regexp.Compile(pattern)
+		if err != nil {
+			klog.ErrorS(err, "Failed to compile regular expression", "pattern", pattern)
+			return sets.Set[string]{}
+		}
+		for _, node := range nodes {
+			if reg.MatchString(node.Name) {
+				members.Insert(node.Name)
+			}
+		}
+	}
+	return members
+}
+
+// exactMatchMember retrieves the member name if the selector is an exact match.
+func (hni *HyperNodesInfo) exactMatchMember(selector topologyv1alpha1.MemberSelector) string {
+	if selector.ExactMatch != nil {
+		return selector.ExactMatch.Name
+	}
+	return ""
+}
+
+func (hni *HyperNodesInfo) updateAncestors(name string) error {
+	ancestors := hni.GetAncestors(name)
```

Comment:
I encountered a bug here:
When I deleted a tier 1 hypernode and then added back later, it didn't set its parent yet (actually it does have its ancestors), but `GetAncestors` only return itself here, which will only reset realNodesSet of itself, not updating its ancestors' realNodesSet

Reply:
That's a good catch! Will fix later.